File: code/python/Queue/dynamic/Queue.py
from Node import Node
from StructBase import StructBase
from Queue.interface import QueueInterface


class Queue(StructBase, QueueInterface):

    def __init__(self) -> None:
        super()
        self._head = None
        self._tail = None

    def __del__(self) -> None:
        self._head = None
        self._tail = None

    def __str__(self) -> str:
        str_ = "\n\tHEAD -> "
        str_ += " -> ".join([str(e) for e in self.__iter__()])
        str_ += " <- TAIL\n"
        return str_

    def __iter__(self) -> "Queue":
        self._tmpNode = self._head
        return self

    def __next__(self) -> int:
        if self._tmpNode is None:
            raise StopIteration
        data = self._tmpNode
        self._tmpNode = data.link
        return data

    def is_empty(self) -> bool:
        return self._head is None and self._tail is None

    def is_full(self) -> bool:
        return False

    def enqueue(self, value: Node):
        if self.is_empty():
            self._head = value
            self._tail = value
            return
        self._tail.link = value
        self._tail = value

    def dequeue(self) -> Node:
        try:
            if self.is_empty():
                raise Exception("Queue is empty.")
            value = self._head
            self._head = value.link
            value.link = None
            return value
        except Exception as exp:
            msg = exp.args
            print(f"Error: {msg[0]}")

    def head(self) -> Node:
        return self._head

    def tail(self) -> Node:
        return self._tail

    @staticmethod
    def options():
        return [
            "Enqueue",
            "Dequeue",
            "Show head",
            "Show tail",
            "Print queue",
            "Exit",
        ]

File: code/typescript/Queue/dynamic/Queue.ts
import { Node } from "../../Node/Node";
import StructBase from "../../StructBase/StructBase";
import QueueInterface from "../interface/QueueInterface";

export default class Queue extends StructBase implements QueueInterface {
  private _head: Node | undefined;
  private _tail: Node | undefined;

  constructor() {
    super();
  }

  *[Symbol.iterator]() {
    let tmpNode = this._head;
    let auxNode = this._head;
    while (tmpNode !== undefined) {
      auxNode = tmpNode;
      tmpNode = tmpNode.link;
      yield auxNode;
    }
  }

  isEmpty(): boolean {
    return !this._head && !this._tail;
  }

  isFull(): boolean {
    return false;
  }

  enqueue(value: Node) {
    try {
      if (this.isEmpty()) {
        this._head = value;
        this._tail = value;
        return;
      }
      if (!this._tail) return;
      value.link = this._tail;
      this._tail.link = value;
    } catch (err) {
      console.error(err);
    }
  }

  dequeue(): Node | undefined {
    try {
      if (this.isEmpty()) throw "Queue is empty.";
      if (!this._head) return;
      const node: Node = this._head;
      this._head = node.link;
      return node;
    } catch (err: any) {
      console.error(err);
    }
  }

  head(): Node | undefined {
    if (this.isEmpty()) return undefined;
    return this._head;
  }

  tail(): Node | undefined {
    if (this.isEmpty()) return undefined;
    return this._tail;
  }

  toString(): string {
    let str: string = "\nFRONT -> ";
    let aux: Node | undefined = this._head;
    str += [...this].map((elem) => elem.data).join(" -> ");
    str += `${aux?.data} <- TAIL\n`;
    return str;
  }
}
